### ISSUE TITLE
ugrep can't read files with permissions 000

### ISSUE THREAD

**Issue by dingus9** [Genivia/ugrep#447]
I encountered a small issue where ugrep fails with the error `ugrep: warning: cannot read tmp.txt: Permission denied` and thought it interesting enough to see if you know about it or have a workaround etc. As root this shouldn't be an issue.

I encountered this on a mounted raw RHEL VM disk image specifically searching /etc/shadow- where the permissions bits are set to 000. VM image is XFS but I am able to replicate this in an ubuntu based docker container image under ext4.

This does not present when using ggrep or cat, only ugrep.

Replication example:
```
root@19f30302819b:/# echo test >tmp.txt 
root@19f30302819b:/# chmod 000 tmp.txt 
root@19f30302819b:/# ll tmp.txt 
---------- 1 root root 5 Dec 24 18:48 tmp.txt
root@19f30302819b:/# cat tmp.txt 
test
root@19f30302819b:/# grep -F test tmp.txt 
test
root@19f30302819b:/# ugrep -F test tmp.txt 
ugrep: warning: cannot read tmp.txt: Permission denied
root@19f30302819b:/# stat tmp.txt 
  File: tmp.txt
  Size: 5         	Blocks: 8          IO Block: 4096   regular file
Device: 36h/54d	Inode: 937209      Links: 1
Access: (0000/----------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-12-24 18:48:34.091147724 +0000
Modify: 2024-12-24 18:48:17.039014549 +0000
Change: 2024-12-24 18:48:23.559065443 +0000
 Birth: 2024-12-24 15:13:51.461371331 +0000

```

Looking at an strace it appears that ugrep aborts after the newfstatat call... does it try to read the file or just says oh no probably not? Root doesn't have to follow the rules. Maybe ugrep is trying to be to good without at least trying to break things a little. Comparing the strace with ggrep. It looks like ggrep just makes a call to openat and doesn't care about stat.

ugrep strace snippet
```
ioctl(2, TCGETS, {B38400 opost isig icanon echo ...}) = 0
newfstatat(AT_FDCWD, "/tmp.txt", {st_mode=S_IFREG|000, st_size=5, ...}, 0) = 0
write(2, "\33[mugrep: \33[1;35mwarning:\33[m \33[1"..., 86ugrep: warning: cannot read /tmp.txt: Permission denied
) = 86
exit_group(2)         
```
ggrep snippet
```
newfstatat(AT_FDCWD, "/dev/null", {st_mode=S_IFCHR|0666, st_rdev=makedev(0x1, 0x3), ...}, 0) = 0
openat(AT_FDCWD, "/tmp.txt", O_RDONLY|O_NOCTTY) = 3
newfstatat(3, "", {st_mode=S_IFREG|000, st_size=5, ...}, AT_EMPTY_PATH) = 0
read(3, "test\n", 98304)                = 5
newfstatat(1, "", {st_mode=S_IFCHR|0620, st_rdev=makedev(0x88, 0), ...}, AT_EMPTY_PATH) = 0
write(1, "test\n", 5test
)                   = 5
read(3, "", 98304)                      = 0
close(3)                                = 0
close(1)                                = 0
close(2)                                = 0
exit_group(0)  
```

```
ugrep --version
ugrep 6.4.1 x86_64-pc-linux-gnu +avx2; -P:pcre2jit; -z:zlib,bzip2,lzma,lz4,zstd,7z,tar/pax/cpio/zip
```

**Comment by genivia-inc** [comment]
Ugrep uses `stat` to determine if an argument is a file or directory and if it exists. It does this to avoid starting the TUI when the specified file or directory argument does not exist, which would go unnoticed in the TUI. After all, we can't see this error anymore in the TUI, which would be confusing to the user not knowing why matches for file(s) aren't showing up. It performs this check even when the TUI is not used, to check if an argument is a directory. If one of the arguments is a directory, then it runs the worker pool to search files in that directory. So there is more going on than GNU grep.

I need time to think about this to work around this.

**Comment by genivia-inc** [comment]
Let me explain that the reason I added this `stat` check is for the TUI. If the specified files cannot be searched then the TUI shows nothing at all, but a nonzero indication of a warning in the status line. From a user perspective, this is probably very confusing why the file does not show up.

So the question is, should we remove the `stat` + readability check when we don't use the TUI? It resolves the issue, but this still means that TUI can't be used in your scenario.

**Comment by dingus9** [comment]
Maybe there's a way to differentiate between stat and real(effective) read/execute access? It's a weird artifact of *nix though that root can do things even though the perms aren't set to allow owner access. For example root id 0 can always write to files even when a file doesn't have write bits set. 

My temporary work around is to have the script I'm running save the octal perms, chmod u+r and then restore the perms after ugrep reads the file. In an entire disk image ugrep throws this error four times. Once for /etc/shadow and shadow- and then again for /etc/group and group-.

**Comment by genivia-inc** [comment]
I agree that the current behavior is not ideal for su scenarios, like yours. I'm thinking about removing the `stat` readability check, which is not something that ugrep users rely on anyway. The minor possibility of confusion with the TUI is not a big price to pay. There is a nonzero warning shown in the status line when one or more files aren't readable (when executing as non-su), so it's not that users aren't warned that one or more files aren't searched with the TUI.

**Comment by genivia-inc** [comment]
Fixed in 7.1.2.
